### ISSUE TITLE
<details> open attribute not synchronised

### ISSUE THREAD

**Issue by afenton90** [facebook/react#15486]
**What is the current behavior?**
`open` attribute does not synchronise with the virtual dom on the `details` element, when using the `details` element as a controlled component. 

I have a codesandbox [here](https://codesandbox.io/s/xl756mk82w) that demonstrates the current behaviour. 

**What is the expected behavior?**
The `open` attribute stays synchronised with the virtual dom and is added/removed.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

React: **`16.8.6`**
Browser: **Chrome**
Did this work in previous versions of React? **No**

**Comment by baruchvlz** [comment]
It looks like it does synchronizes with the virtual dom.

![image](https://user-images.githubusercontent.com/14321495/56666916-18070780-66ad-11e9-8e60-4cdde0c1c20a.png)


Also, you don't need to do `open={open}` for the details element. HTML understands that clicking on the `<summary>` toggles the details. You can keep the `onClick={onToggle}` too. Check it [here](https://codesandbox.io/s/wqppk4k445)

**Comment by afenton90** [comment]
Thanks @baruchvlz. So is the official advice that `details` cannot be used as controlled component?

**Comment by alejandronanez** [comment]
I just ran into the same issue today, and I fixed it by adding an `event.preventDefault` to the `onClick` handler. It is not crystal clear to me why this works tbh.

```js
const onToggle = event => {
  event.preventDefault();
  setOpen(!open);
};

// Later on...
return (
  <div className="App">
    <details open={open} onClick={onToggle}>...</details>
  </div>
)
```

[Codesandbox](https://codesandbox.io/s/jvzky8yz23)

**Comment by afenton90** [comment]
Awesome @alejandronanez I'll go with that for now.
I'll let the maintainers decide whether this issue is a bug or just a quirk of the `details` element.

**Comment by zgoby** [comment]
@alejandronanez is right!details element have Default behavior! So,it is not controlled by onclickEvent

**Comment by julienw** [comment]
Also the `<details>` element supports a `toggle` event (`onToggle` with React) that could make it possible to keep the native implementation while still updating the local state.

**Comment by anadutova** [comment]
`Event.preventDefault()` breaks click events of children, e.g. `<select>`

The `ontoggle` attribute is not standard and as such is not included in type definitions.

My issue is that I want close the `details` element when it is clicked outside of it. I also append a class depending on whether it is open.

The controlled component approach breaks because -- I assume -- both the state update and the native implementation are executed, causing it to toggle twice in a row. The `open` attribute is not synchronised.

My current approach is to call `setAttribute` on the `details` DOM node. I find this clumsy compared to using a controlled component.

**Comment by julienw** [comment]
This "stale" bot thing is painful. It shouldn't close bugs without someone checks it doesn't happen anymore...

**Comment by wesleyboar** [comment]
I do not have a solution for `onClick`, but I believe I have noticed a common discrepancy, in this thread, among attempts to mimic `<details>` behavior with React, and actual `<details>` behavior.

**Summary**
The live code samples provided thus far set `onClick` on the `<details>` element. This does not accurately represent the standard behavior of a `<details>` tag.

**Explanation**
In a raw `<details>/<summary>`, if one clicks on the content within an open `<details>`, underneath the `<summary>`, the `<details>` element does **not** close—only a click (or other activating user action) on the `<summary>` element will close (and open) the `<details>`. But, in the live code samples provided thus far, clicking on the described region closes the `<details>` element.

**Example**
Without React:
<details>
<summary>Summary</summary>
<p>Content (clicking here has no effect)</p>
</details>

With React:
https://codesandbox.io/s/react-details-bug-ov9sl

**Reference**
- The [`<details>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) content is shown on "click" of the `<summary>` tag.
- The [`toggle` event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDetailsElement/toggle_event) (not `click` event) is the `<details>` element event for changing the `open` state.

**Browser Support**
[Chrome and Edge only recently have support](https://caniuse.com/#feat=mdn-api_htmldetailselement_toggle_event)

**Comment by bitworking** [comment]
To mimic the default behavior I did the following:

```
<details open={open} onClick={onToggle}>
  <summary>{title}</summary>
  <div role="region" onClick={(e) => e.stopPropagation()}>
    {children}
  </div>
</details>
```

**Comment by butchler** [comment]
The title of this issue is inaccurate: React does keep the `open` attribute in sync with the DOM (as mentioned above https://github.com/facebook/react/issues/15486#issuecomment-486263985). However, this is still a confusing case even though React isn't technically doing anything wrong.

It's confusing because, unlike most other native DOM elements, the `<details>` element has state which it can update by itself without React knowing about it. This is what happens in the codesandbox from the issue description (https://codesandbox.io/s/xl756mk82w) when you click on the element for the first time:

1. User clicks on `<details>` element
2. The `onClick` handler is triggered, and the React state is toggled from false to true
3. React (when not in concurrent mode) immediately re-renders, setting the `open` attribute of the `<details>` element to true
4. The default behavior of the `<details>` element kicks in, toggling the `open` state, reversing the `open` attribute back to false (but React doesn't know about this)

This is why the element does not open the first time you click. Because this happens instantly, it's not obvious that the state is actually being toggled twice. When you click on the element again, this happens:

1. User clicks on `<details>` element
2. The `onClick` handler is triggered, and React still thinks the element is open, so the React state is toggled from true to false
3. React re-renders, but the open attribute is already false (because the native behavior reversed what React did in the first click), so React does not update the DOM
4. The default behavior kicks in, and the element is toggled

This is why you can toggle the element after the first click, even though React and the DOM disagree about the current state. Every time it re-renders, React sees that the DOM is already in the state that it wants to be in, so it does not update the DOM and behaves as if the `<details>` element was an uncontrolled compoent (even though it is technically being controlled).

As mentioned in https://github.com/facebook/react/issues/15486#issuecomment-488028431, the solution is to call `event.preventDefault()` in the `onClick` handler to disable the browser's default behavior so that React and only React controls the state. Note that as https://github.com/facebook/react/issues/15486#issuecomment-669674869 mentions, to mimic native behavior the `onClick` handler should be on the `<summary>` element, not the `<details>` element.

https://github.com/facebook/react/issues/15486#issuecomment-525324389 mentioned that there is a `toggle` event which can be used to notify React about the current state. This might be useful when using the `<details>` element as a non-controlled component, but is a fragile way to use it as a controlled component because you still have the problem of the native behavior toggling the state separately from React. This codesandbox demonstrates how this could be an issue: https://codesandbox.io/s/react-details-bug-forked-iucdn. Clicking on the `<details>` element itself works correctly, but it causes an infinite re-rendering loop when you click on the button that toggles the React state alone. This is because `onToggle` is trickered not only when the user interacts with the element, but also triggered when React itself modifies the `open` attribute. This means that every time React toggles `open`, `onToggle` is triggered, making React toggle `open`, making `onToggle` get triggered again, etc.

**tl;dr** This is a confusing case, but is probably not an issue with React. The solution is to call `event.preventDefault()` in the `onClick` handler of the `<summary>` element.

**Comment by campbellgoe** [comment]
Finally, as the above comments found a solution: to use event.preventDefault() inside an onClick on the summary element, here is the code.
You must also specify the open attribute on the details element, and have no open attribute not even `open={false}` and therefore use `{...(open ? { open: true } : {})}`.

```jsx
import { useState } from "react";

export default function Details({ summary, children, startOpen = false }){
  const [open, setOpen] = useState(startOpen);
  return (
    <details {...(open ? { open: true } : {})}>
      <summary
        onClick={(e) => {
          e.preventDefault();
          setOpen((open) => !open);
        }}
      >
        {typeof summary === "function" ? summary(open) : summary}
      </summary>
      {open && children}
    </details>
  );
};

```
Usage:
```jsx
function App() {
  return (
    <div className="App">
      <Details summary={(open) => `${open ? "View" : "Hide"} contents`}>
        <p>The summary prop can be a function which gets the open value</p>
      </Details>
      <hr/>
      <Details summary="Details" startOpen={true}>
        <p>
          These contents are added/removed from the DOM depending on whether
          details is open or not.
        </p>
      </Details>
      <hr/>
    </div>
  );
}
```

See working / edit the code: https://codesandbox.io/s/beautiful-borg-nbh93?file=/src/App.js

**Comment by iampava** [comment]
I just stumbled a somewhat similar bug, that I think is related to this.

I have a `<details>` element that is open by default, and the entire page flickers. Here's a CodeSandbox: https://codesandbox.io/s/throbbing-grass-1tphyl?file=/src/App.js

**Comment by butchler** [comment]
> I have a `<details>` element that is open by default, and the entire page flickers. Here's a CodeSandbox: https://codesandbox.io/s/throbbing-grass-1tphyl?file=/src/App.js

The explanation was hidden in a large comment, but I mentioned this case above: https://github.com/facebook/react/issues/15486#issuecomment-873516817

> [#15486 (comment)](https://github.com/facebook/react/issues/15486#issuecomment-525324389) mentioned that there is a `toggle` event which can be used to notify React about the current state. This might be useful when using the `<details>` element as a non-controlled component, but is a fragile way to use it as a controlled component because you still have the problem of the native behavior toggling the state separately from React. This codesandbox demonstrates how this could be an issue: https://codesandbox.io/s/react-details-bug-forked-iucdn. Clicking on the `<details>` element itself works correctly, but it causes an infinite re-rendering loop when you click on the button that toggles the React state alone. This is because `onToggle` is triggered not only when the user interacts with the element, but also triggered when React itself modifies the `open` attribute. This means that every time React toggles `open`, `onToggle` is triggered, making React toggle `open`, making `onToggle` get triggered again, etc.

It seems like my sandbox did not include the button that I mentioned in the comment, but I updated it now to add the button.

In your sandbox, there is no button for toggling the React state alone, but because `isOpen` defaults to true, on mount React will set the `details` element's `open` property to true. This will trigger the `onToggle` event, which will call the event handler, which will flip the value of `isOpen`, which will again cause React to update the `open` property, and that repeats in an infinite loop.

**Comment by iampava** [comment]
Oh, I see. So this is how it works 🤔

Thanks for the explanation. I'm gonna go with the solution proposed above, using a click event listener and `.preventDefault()`.

**Comment by boldium-eric-mikkelsen** [comment]
The idea that breaking an html element isn't a bug, is a problem. If a native element with w3c settled behavior is broken by React, React is broken, and this is a legit bug.

**Comment by butchler** [comment]
As far as I can tell nobody in this thread has suggested that this is not a problem. Maybe you were confused by my earlier comment where I said this?:

> The title of this issue is inaccurate: React does keep the `open` attribute in sync with the DOM (as mentioned above [#15486 (comment)](https://github.com/facebook/react/issues/15486#issuecomment-486263985)).

I simply meant that the specific claim that the title made is technically inaccurate. I do think this is a very confusing thing that will waste a lot of people's time if it is not somehow fixed (or at least better documented).

In case you missed it, it is still possible to use the `details` element as long as you call `event.preventDefault()` in the event handler. Hopefully the React team can find a way to make it possible to use the `details` element without needing to do this, but for now you can use this as a workaround.

**Comment by jsejcksn** [comment]
It's possible to react to the native [`toggle`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDetailsElement/toggle_event) event that's dispatched on the [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) element, using the nested [`open`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDetailsElement/open) property to update state. This feels more idiomatic and avoids the previously mentioned workarounds related to the inner [`<summary>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/summary) element. Here's a minimal, typed example:

[TS Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&removeComments=true&target=99&jsx=4&exactOptionalPropertyTypes=true&inlineSourceMap=true&inlineSources=true&isolatedModules=true&noImplicitOverride=true#code/JYWwDg9gTgLgBAbzjAnmApnASughgYxgFEAbdEdAOxgBo4BXAZ3QGUZcZMBfOAMyggg4AIih5CwgNwAoab3qVCwCJTgBhQZEpUYACgCUALmzjiZCtUTS4cfCsbwA2hAyU6zGAHlXAXTgBeBmY2DnRdXlwSZn0Za1t7eAALXEoAEzIAFQgAc2yyALhdOJt0ADcdYyR8eigxagzcKGz0GEq4FypjACMICDIUuB4uGjj9AIA+OA9vKl0ynQA6atqdBqaWhY7KGNkbMRga1SKbGwAeVJbcYCj2yizcsn8EZLTMnLz0Hi2nra5x4pOp0Y9BAIEaKHGCC2cAA-CIABLAC7CODGYQsRIQADuwi4pwA9MDQeD-idAWBxgBZaCYOzUHQEikAgkXdjXRikuA7LjSIA)

```ts
import { type ReactElement, useState } from "react";

function Component(): ReactElement {
  const [open, setOpen] = useState(false);

  const handleToggle = (
    event: { currentTarget: { open: boolean } },
  ) => setOpen(event.currentTarget.open);

  return (
    <details onToggle={handleToggle} open={open}>
      <summary>{open ? "Hide" : "Show"}</summary>
      <p>More content</p>
    </details>
  );
}

```

**Comment by butchler** [comment]
Using `onToggle` works for simple cases but may lead to an infinite re-rendering loop if the state is updated by anything other than the event handler. From [my previous comment](https://github.com/facebook/react/issues/15486#issuecomment-873516817):

> [#15486 (comment)](https://github.com/facebook/react/issues/15486#issuecomment-525324389) mentioned that there is a `toggle` event which can be used to notify React about the current state. This might be useful when using the `<details>` element as a non-controlled component, but is a fragile way to use it as a controlled component because you still have the problem of the native behavior toggling the state separately from React. This codesandbox demonstrates how this could be an issue: https://codesandbox.io/s/react-details-bug-forked-iucdn. Clicking on the `<details>` element itself works correctly, but it causes an infinite re-rendering loop when you click on the button that toggles the React state alone. This is because `onToggle` is triggered not only when the user interacts with the element, but also triggered when React itself modifies the `open` attribute. This means that every time React toggles `open`, `onToggle` is triggered, making React toggle `open`, making `onToggle` get triggered again, etc.

**Comment by jsejcksn** [comment]
> Using `onToggle` works for simple cases but may lead to an infinite re-rendering loop if the state is updated by anything other than the event handler

@butchler That's only an issue when using a function that **inverts** the state value (like in your CodeSandbox example) — that's what causes the race condition and infinite re-render scenario. When **setting** it according to the value in the dispatched event (like in the code shown in my previous [comment](https://github.com/facebook/react/issues/15486#issuecomment-1537086764)), you can be confident that won't happen.

If the new boolean value is opposite of the current value, it will re-render once. If it's the same, no new render will occur. From the React [docs](https://react.dev/reference/react/useState#setstate-caveats):

> If the new value you provide is identical to the current `state`, as determined by an [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) comparison, React will **skip re-rendering the component and its children**. This is an optimization. Although in some cases React may still need to call your component before skipping the children, it shouldn’t affect your code.

**Comment by butchler** [comment]
Correct, and that's what I meant by "if the state is updated by anything other than the event handler". Sorry if my phrasing wasn't clear.

This means that for the case where you only need to know if the `detail` element is open or closed in your React component, using `onToggle` works fine. However, in some cases you might want to toggle the `detail` element via React state. For example, you might want to have a "expand all"/"collapse all" button that toggles all of the `detail` elements on a particular page. In that case, using `onToggle` will lead to an infinite render loop, and for most people it won't be obvious what's wrong or how to fix it.

For that reason, personally I wouldn't recommend `onToggle` as the preferred way to use `detail` elements. Ultimately it comes down to personal preference. In my case, I don't mind having to add the `event.preventDefault()` call if it means I never have to think about this potential edge case with `onToggle`.

**Comment by butchler** [comment]
If you do use `onToggle`, I would recommend that you don't pass the `open` prop to the `detail` element. It gives the false impression that you can toggle the element with React state updates alone (i.e. not with the event handler) when in reality that will lead to an infinite render loop.

**Comment by jsejcksn** [comment]
> If you do use `onToggle`, I would recommend that you don't pass the `open` prop to the `detail` element. It gives the false impression that you can toggle the element with React state updates alone (i.e. not with the event handler) when in reality that will lead to an infinite render loop.

@butchler You can have confidence that won't be an issue either — it'll still work!

Below I've included a reproducible example that you can use to verify. No build tools are needed — just save the contents as a local file and open in a browser. In case you don't have a text editor available, here's a link to a hosted version: https://jsfiddle.net/4w7tuLfk/

<details>
<summary><code>example.html</code></summary>

```html
<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <title>Details demo</title>
  <style>body { font-family: sans-serif; } button { margin: 1rem 0; } details { border: 1px solid blue; } summary { border: 1px solid red; }</style>
  <script src="https://cdn.jsdelivr.net/npm/react@18.2.0/umd/react.production.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/react-dom@18.2.0/umd/react-dom.production.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/@babel/standalone@7.21.8/babel.min.js"></script>
  <script type="text/babel" data-type="module" data-presets="env,react">
    // This example uses UMD modules
    const { useRef, useState } = React;

    function RenderCount() {
      const renderCountRef = useRef(0);
      renderCountRef.current += 1;
      return <div>Render count: {renderCountRef.current}</div>;
    }

    function App() {
      const [open, setOpen] = useState(false);

      return (
        <>
          <RenderCount />
          <button onClick={() => setOpen((open) => !open)}>
            Toggle using state
          </button>
          <details
            onToggle={(event) => setOpen(event.currentTarget.open)}
            open={open}
          >
            <summary>{open ? "Hide" : "Show"}</summary>
            <p>More content</p>
          </details>
        </>
      );
    }

    ReactDOM.createRoot(document.getElementById("root")).render(<App />);
  </script>
</head>

<body>
  <div id="root"></div>
</body>

</html>

```

</details>

**Comment by butchler** [comment]
I apologize, I didn't notice the difference between your suggestion and the previous suggestions in this thread about using the `toggle` event. The key difference is to not directly toggle the React state to, but to always set the React state to match the `detail` element's current `.open` state. That does indeed avoid the problem I mentioned. 👍

In hindsight it seems so obvious that I don't know why I never thought of it myself. 😅

**Comment by butchler** [comment]
I noticed that the new React docs have pages documenting how to use common types of controlled elements such as `input`: https://react.dev/reference/react-dom/components/input.

If someone added a similar page describing usage of the `detail` element with React then maybe that would be enough to close this ticket?

**Comment by SheepTester** [comment]
Note that browsers can open `<details>` without firing `click` (but it does fire `toggle`, but you can't `preventDefault` it). If you search for content in the page, browsers can open `<details>` if its contents contain it . If you try to force it closed using `e => (e.currentTarget.open = false)`, it still shows a flash of the contents being open.

Here's a demo: https://jsfiddle.net/SheepTester/57chzoen/

**Comment by lufinkey** [comment]
I just read over this issue, and I'm realizing nobody has brought up the fact that `<summary>` is the element that is meant to be clickable, not `<details>`. I'm guessing this is why preventing the default behavior of the element "fixes" the issue

**Comment by IVLIU** [comment]
wrap setState with startTransition.
![Image](https://github.com/user-attachments/assets/41810fa5-16e7-47c3-816b-37984c59e2a8)
